File: xsync_core/path_utils.py
import os

def process_include_paths(raw_lines, mode, local_dir, remote_dir):
    """
    Cleans paths: strips absolute path prefixes matching either local_dir or remote_dir,
    removes comments, and appends a trailing slash for directories in push mode.
    """
    processed = []
    local_dir = os.path.normpath(local_dir) if local_dir else ""
    remote_dir = os.path.normpath(remote_dir) if remote_dir else ""
    
    for line in raw_lines:
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        # Strip quotes (terminal sometimes adds them when dragging files)
        if (line.startswith('"') and line.endswith('"')) or (line.startswith("'") and line.endswith("'")):
            line = line[1:-1].strip()
            
        norm_line = os.path.normpath(line)
        
        # Check remote_dir prefix
        if remote_dir and (norm_line.startswith(remote_dir + os.sep) or norm_line.startswith(remote_dir.replace("\\", "/") + "/")):
            rel = norm_line[len(remote_dir) + 1:]
        elif remote_dir and norm_line == remote_dir:
            rel = "."
        # Check local_dir prefix
        elif local_dir and (norm_line.startswith(local_dir + os.sep) or norm_line.startswith(local_dir.replace("\\", "/") + "/")):
            rel = norm_line[len(local_dir) + 1:]
        elif local_dir and norm_line == local_dir:
            rel = "."
        else:
            rel = line
            
        # Clean leading slashes or dots from relative path
        rel = rel.lstrip("/").lstrip("\\")
        
        # If it is a directory locally and does not end with /, append /
        # This only applies to push mode as we check local filesystems
        if mode == "push" and local_dir:
            full_path = os.path.join(local_dir, rel)
            if not rel.endswith("/") and os.path.isdir(full_path):
                rel = rel + "/"
                
        processed.append(rel)
    return processed

def build_include_filter(include_lines):
    """
    Constructs an rsync merge filter list from whitelist paths.
    For each path 'a/b/c':
      - adds '+ a/'
      - adds '+ a/b/'
      - adds '+ a/b/c' or '+ a/b/c/**' (if directory)
    Finally appends '- *' to exclude everything else.
    """
    filters = []
    roots = set()
    
    for line in include_lines:
        line = line.strip()
        if not line or line.startswith("#"):
            continue
            
        # Strip leading '+' or '-' prefix if user typed it
        if line.startswith("+ ") or line.startswith("- "):
            line = line[2:].strip()
            
        # Strip leading ./
        if line.startswith("./"):
            line = line[2:]
            
        # Normalize double slashes
        while "//" in line:
            line = line.replace("//", "/")
            
        if line.startswith("/"):
            # Ignore absolute paths
            continue
            
        is_dir = False
        if line.endswith("/"):
            is_dir = True
            line = line[:-1]
            
        if not line:
            continue
            
        parts = [p for p in line.split("/") if p]
        if not parts:
            continue
            
        # Add parent directories
        current = ""
        for p in parts[:-1]:
            current = f"{current}/{p}" if current else p
            filters.append(f"+ {current}/")
            
        # Add target path
        if is_dir:
            filters.append(f"+ {line}/")
            filters.append(f"+ {line}/**")
        else:
            filters.append(f"+ {line}")
            
        roots.add(parts[0])
        
    if not roots:
        return None
        
    # Deduplicate while preserving order
    seen = set()
    deduped_filters = []
    for f in filters:
        if f not in seen:
            seen.add(f)
            deduped_filters.append(f)
            
    deduped_filters.append("- *")
    return deduped_filters

File: internal/pathutils/pathutils.go
package pathutils

import (
	"os"
	"path/filepath"
	"strings"
)

// ProcessIncludePaths cleans paths: strips absolute prefixes matching either localDir or remoteDir,
// removes comments, and appends a trailing slash for directories in push mode.
func ProcessIncludePaths(rawLines []string, mode, localDir, remoteDir string) []string {
	var processed []string

	cleanLocal := ""
	if localDir != "" {
		cleanLocal = filepath.Clean(localDir)
	}
	cleanRemote := ""
	if remoteDir != "" {
		cleanRemote = filepath.Clean(remoteDir)
	}

	for _, line := range rawLines {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}

		// Strip quotes
		if (strings.HasPrefix(line, "\"") && strings.HasSuffix(line, "\"")) ||
			(strings.HasPrefix(line, "'") && strings.HasSuffix(line, "'")) {
			line = strings.TrimSpace(line[1 : len(line)-1])
		}

		normLine := filepath.Clean(line)
		rel := line

		// Check remoteDir prefix
		if cleanRemote != "" {
			if strings.HasPrefix(normLine, cleanRemote+string(os.PathSeparator)) ||
				strings.HasPrefix(normLine, strings.ReplaceAll(cleanRemote, "\\", "/")+"/") {
				rel = normLine[len(cleanRemote)+1:]
			} else if normLine == cleanRemote {
				rel = "."
			}
		}

		// Check localDir prefix
		if cleanLocal != "" {
			if strings.HasPrefix(normLine, cleanLocal+string(os.PathSeparator)) ||
				strings.HasPrefix(normLine, strings.ReplaceAll(cleanLocal, "\\", "/")+"/") {
				rel = normLine[len(cleanLocal)+1:]
			} else if normLine == cleanLocal {
				rel = "."
			}
		}

		// Clean leading slashes
		rel = strings.TrimLeft(rel, "/\\")

		// If it is a directory locally and does not end with /, append / in push mode
		if mode == "push" && cleanLocal != "" {
			fullPath := filepath.Join(cleanLocal, rel)
			if info, err := os.Stat(fullPath); err == nil && info.IsDir() {
				if !strings.HasSuffix(rel, "/") {
					rel = rel + "/"
				}
			}
		}

		if rel != "" {
			processed = append(processed, rel)
		}
	}

	return processed
}

// BuildIncludeFilter constructs an rsync merge filter list from whitelist paths.
// For each path 'a/b/c':
//   - adds '+ a/'
//   - adds '+ a/b/'
//   - adds '+ a/b/c' or '+ a/b/c/**' (if directory)
// Finally appends '- *' to exclude everything else.
func BuildIncludeFilter(includeLines []string) []string {
	var filters []string
	roots := make(map[string]bool)

	for _, line := range includeLines {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}

		// Strip leading + or -
		if strings.HasPrefix(line, "+ ") || strings.HasPrefix(line, "- ") {
			line = strings.TrimSpace(line[2:])
		}

		// Strip leading ./
		if strings.HasPrefix(line, "./") {
			line = line[2:]
		}

		// Normalize double slashes
		for strings.Contains(line, "//") {
			line = strings.ReplaceAll(line, "//", "/")
		}

		if strings.HasPrefix(line, "/") {
			// Ignore absolute paths
			continue
		}

		isDir := false
		if strings.HasSuffix(line, "/") {
			isDir = true
			line = strings.TrimSuffix(line, "/")
		}

		if line == "" {
			continue
		}

		parts := strings.Split(line, "/")
		var cleanParts []string
		for _, p := range parts {
			if p != "" {
				cleanParts = append(cleanParts, p)
			}
		}
		if len(cleanParts) == 0 {
			continue
		}

		// Add parent directories
		current := ""
		for i := 0; i < len(cleanParts)-1; i++ {
			p := cleanParts[i]
			if current != "" {
				current = current + "/" + p
			} else {
				current = p
			}
			filters = append(filters, "+ "+current+"/")
		}

		// Add target path
		if isDir {
			filters = append(filters, "+ "+line+"/")
			filters = append(filters, "+ "+line+"/**")
		} else {
			filters = append(filters, "+ "+line)
		}

		roots[cleanParts[0]] = true
	}

	if len(roots) == 0 {
		return nil
	}

	// Deduplicate preserving order
	seen := make(map[string]bool)
	var deduped []string
	for _, f := range filters {
		if !seen[f] {
			seen[f] = true
			deduped = append(deduped, f)
		}
	}

	deduped = append(deduped, "- *")
	return deduped
}
